File: data-structures/linked-list/python/linked_list.py
"""
Linked List data structure implementation
"""

# Author: 
# Lajos Neto <lajosnetogit@gmail.com>


class _Node:
    """
    Simple node used for representing values 
    inside the linked list.
    """
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    """
    Linked List main class where all value nodes
    are stored within.
    """
    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0
    
    def __iter__(self):
        iterator = self._head
        while True:
            if iterator is None:
                return
            yield iterator.value
            iterator = iterator.next
    
    def empty(self):
        """Check if the list is empty"""
        return self._size == 0
    
    def insert(self, index, value):
        """Insert a vulue at a given index position"""
        new_node = _Node(value)
        if self.empty() or index == 0:
            self.push_front(value)
        elif index >= self._size or index < 0:
            raise IndexError
        else:
            aux_pointer = self._head
            for i in range(index-1):
                aux_pointer = aux_pointer.next
            new_node.next = aux_pointer.next
            aux_pointer.next = new_node
            self._size += 1
    
    def front(self):
        """Returns the first value on the list"""
        return self._head.value
    
    def push_front(self, value):
        """Insert a value at the beginning of the list"""
        new_node = _Node(value)
        if self.empty():
            self._head = new_node
            self._tail = new_node
        else:
            new_node.next = self._head
            self._head = new_node
        self._size += 1

File: data-structures/linked-list/kotlin/LinkedList.kt
/**
 * Linked List implementation.
 * 
 * Author : Lajos Neto.
 * Email : lajosnetogit@gmail.com
 */

data class Node<T>(
    val value: T,
    var next: Node<T>? = null
)

class LinkedList<T> {
    private var size = 0
    private var head: Node<T>? = null
    private var tail: Node<T>? = null

    override fun toString(): String {
        var stringRep = ""
        var walker = head
        while (walker != null) {
            stringRep += "${walker.value} -> "
            walker = walker.next
        }
        return "${stringRep}/null"
    }

    fun addBegin(value: T) {
        if (isEmpty()) initialLoad(value)
        else {
            val newNode = Node(value, null)
            newNode.next = head
            head = newNode
        }
        size += 1
    }

    fun add(value: T, index: Int) {
        if (isEmpty() || index == 0) addBegin(value)
        else if (index in 0 until size()) {
            val newNode = Node(value)
            var walker = head
            for (i in 0 until index - 1) walker = walker?.next
            newNode.next = walker?.next
            walker?.next = newNode
            size += 1
        }
    }

    fun removeBegin() {
        if (!isEmpty()) {
            if (size() == 1) {
                head = null
                tail = null
            } else {
                head = head?.next
                size -= 1
            }
        }
    }

    fun addEnd(value: T) {
        if (isEmpty()) initialLoad(value)
        else {
            val newNode = Node(value, null)
            tail?.next = newNode
            tail = newNode
        }
        size += 1
    }

    fun removeEnd() {
        if (!isEmpty()) {
            if (size() == 1) {
                head = null
                tail = null
            } else {
                var walker = head
                while (walker?.next?.next != null) walker = walker.next
                tail = walker
                tail?.next = null
            }
            size -= 1
        }
    }

    fun front() = this.head?.value
    fun back() = this.tail?.value
    fun print() = this.toString()
    fun size() = size
    private fun isEmpty() = size == 0

    private fun initialLoad(value: T) {
        val newNode = Node(value)
        head = newNode
        tail = newNode
        size += 1
    }
}